### PR TITLE
U4-6573 - allow List View to Show Name of Picked Nodes

### PR DESCRIPTION
**Pull request by AndyButland**
This PR will update the list view display of picked nodes that currently rendering the node Id(s) with the appropriate node names.  Implemented for the core content picker and multi-node content picker.

### CODE REVIEW

File: src/Umbraco.Web.UI.Client/src/views/propertyeditors/listview/listview.controller.js [hunk 1]
```diff
@@ -261,11 +261,23 @@ function listViewController($rootScope, $scope, $routeParams, $injector, $cookie
          $scope.actionInProgress = false;
          $scope.listViewResultSet = data;
 
-         //update all values for display
+         // Update all values for display
          if ($scope.listViewResultSet.items) {
-            _.each($scope.listViewResultSet.items, function (e, index) {
-               setPropertyValues(e);
+
+            // For content, track any values that are referended node Ids
```
Comment (Nicholas-Westby) [inline src/Umbraco.Web.UI.Client/src/views/propertyeditors/listview/listview.controller.js]: "referended"?

**Comment by AndyButland** [comment]
Have extended this PR now to also cover [U4-6405](http://issues.umbraco.org/issue/U4-6405), so as to display labels instead of ids for pre-values, when using the core list types.

**Comment by dampee** [comment]
👍  What do we need to do to get this PR merged in?  :-D

**Comment by Shazwazza** [comment]
I've tagged the issue to be prioritized we'll review next sprint

**Comment by Shazwazza** [comment]
I've had a quick skim of the code, my first concern is performance overhead. Having the ability to sort on custom property data is already a massive performance hog and is causing problems on large sites. In fact we need to disable sorting on NText columns since it kills the SQL server http://issues.umbraco.org/issue/U4-9179, second if you have tens of thousands of records and are sorting on a custom property value (even when it is not NText), this has a severe degradation of performance.

So I am certainly worried about adding even more performance impacts here. 

I would also require a lot of code comments for each of these methods since at a glance I have no idea what any of this code is doing.

**Comment by AndyButland** [comment]
@Shazwazza - have added further comments to hopefully clarify a few things and make it easier to review, but will also explain a bit further here.  The implementation - all in the client-side list view controller - is doing the following:

- Checks the columns being rendered in the list view and for any that are from content pickers (and so currently displaying just Ids), creates an array of all the unique content Ids displayed on the current page of the list
- Similarly for columns being rendered from lists (and so currently displaying just pre-value Ids), creates an array of all the unique pre-value Ids
- Does a look-up of the content nodes that match the Ids (causing one extra HTTP request)
- Does a look-up of the pre-values that match the Ids (causing one HTTP request for each data type displayed on the current page of the list)
- Finally with those two collections, iterates over the currently displayed page on the list view and replaces all the Ids with the more appropriate name of the content or label of the pre-value

Concern about performance is definitely valid, but it's not the same thing going on here as with the sorting.  The sorting obviously has to do so over the full set of data being presented in the list view, and so is going to be less and less performant the bigger the data set is (and particularly if inappropriate columns are added to the list view).  As this is client-side, it's making additional requests, but only on the basis of the current page of the list view.  So I believe whether there's one page, or a thousand pages, this wouldn't make a difference.  The concern if any should perhaps be more around the additional HTTP requests being made, but my hope in making the PR of course is that it would be viewed that the benefit of the improved feature outweigh this cost.

**Comment by Shazwazza** [comment]
So yes performance is a concern for 2 reasons:

* 2 extra HTTP requests for every page viewed
* 1 potentially extra large database query to return all IContent objects by a number of IDs
* n+1 database queries to lookup pre-values for every row

Combine this with large data sets and sorting on a custom column (even if it's not NText) and paging is going to start being really slow and maybe worse is the DB server is going to be spiking. 

My main concern is that the whole performance of the list view is becoming slower and slower. Will have to think about this.

**Comment by AndyButland** [comment]
It's not quite as bad as that:

- Those 2 extra HTTP requests would only be if the list view has both picked content and list based data types.   If those types of data types aren't used in the column list, the requests aren't made.
- The content query will only get IDs for the uniquely picked content items on the currently viewed page of the list view, so that would be unlikely to be more than a handful
- It's not N+1, rather one request per list based data type.  So if you had one column based on a drop down list and one on a radio button, that would be two additional requests.

But I take the general point and can see you'll need to consider whether the benefits outweigh the potential performance issues.  I wonder though if it may be something that will be useful and not harmful in the majority of cases, but if used "irresponsibly" - i.e. creating a list view on a massive data set, with several custom columns including content pickers and list data types - then you might find issues.  The question then I guess is whether removing the risk of people shooting themselves in the foot like in these latter cases has to mean the feature of using these types of columns usefully on the list view shouldn't be implemented at all.

**Comment by AndyButland** [comment]
Just explaining the above further commit on this PR - to restrict ordering on custom columns only to those appropriate for sorting.  Partly related to the performance issues discussed above.

But also as sorting will be based on the raw value, so if this PR was included such that a page name rather than an id is displayed for pages picked via a content picker, sorting would still use the ID and hence the list would not be sorted by the page name as the user would expect.

**Comment by benbracedigital** [comment]
Is this PR close to being merged in @Shazwazza with Andy's latest improvements? It would be a great to have this working as node IDs don't mean anything to editors.
@AndyButland does this work for MNTP as well as content pickers?

**Comment by Shazwazza** [comment]
This requires a lot of testing for performance because we've fixed quite a lot of things with list view performance recently and although there are claims that it won't affect it too much, I have my reservations. It needs to be tested with large data sets and we need to actually understand the real perf implications, not just by assuming it's fine because it works fast locally. 

This is currently not scheduled but we will have a look at it in due time.

**Comment by AndyButland** [comment]
@benbracedigital - yes, MNTP too.

**Comment by AndyButland** [comment]
@madsrasmussen - just flagging you on this PR following our brief chat about it following your talk at Codegarden

**Comment by nul800sebastiaan** [comment]
Hey there @AndyButland!

I'm writing today after anouncing at [Codegarden](https://codegarden19.com/), our yearly conference, that the upcoming version 7.15.0 will be the last Umbraco v7 release with new features in it. From now on all our efforts will be focused on version 8 instead. 🎱

We're wrapping up the 7.15 version at the moment and unfortunately these proposed changes won't be able to make it in there. 

We'd like to extend a big thank you the efforts on this so far and we would love it if you could have a look at porting this feature over to v8 instead. Of course if you have any questions on where to start then feel free to [ask on the forum](https://our.umbraco.com/forum/contributing-to-umbraco-cms/), we monitor all questions there and we're happy to assist.

Thanks again for the work in this PR and we hope to see you contributing to v8 some time soon!

Best, Sebastiaan and the Umbraco CMS team.

**Pull request closed without merging**